File: moling-server/app/router/project_health.py
"""墨灵 (Moling) — Project Health API Router.

提供项目级健康告警端点：
- GET /{project_id}/health - 获取项目健康告警
- POST /{project_id}/health/refresh - 刷新项目健康告警
"""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from datetime import datetime, timezone

from app.dependencies import get_db, get_current_user
from app.schemas.health import HealthCheckResp, HealthAlertItem
from app.service import health_service
from app.errors import NotFoundError, ErrorCode, AppError

router = APIRouter(tags=["project-health"])


@router.get("/{project_id}/health", response_model=HealthCheckResp)
async def get_project_health(
    project_id: int,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> HealthCheckResp:
    """获取项目健康告警。

    返回项目的健康告警列表（R1/R2/R3规则检查结果）。

    Args:
        project_id: 项目 ID
        db: 数据库会话
        current_user: 当前认证用户

    Returns:
        HealthCheckResp: 包含告警列表和检查时间
    """
    # 验证项目存在且属于当前用户
    from app.dao import project_dao
    project = await project_dao.get(db, project_id)
    if project is None:
        raise NotFoundError(
            error_code=ErrorCode.PROJECT_NOT_FOUND,
            detail="Project not found",
        )
    if project.user_id != current_user.id:
        raise AppError(
            error_code=ErrorCode.PROJECT_ACCESS_DENIED,
            detail="Not authorized to access this project",
        )

    # 获取活跃告警
    alerts = await health_service.get_alerts(db, project_id, active_only=True)

    # 转换为响应格式
    alert_items = []
    latest_checked_at = None
    for alert in alerts:
        alert_items.append(HealthAlertItem(
            rule=alert.rule,
            title=alert.title,
            detail=alert.detail,
        ))
        if alert.checked_at:
            if latest_checked_at is None or alert.checked_at > latest_checked_at:
                latest_checked_at = alert.checked_at

    # 如果没有告警或没有检查时间，使用当前时间
    if latest_checked_at is None:
        latest_checked_at = datetime.now(timezone.utc)

    return HealthCheckResp(
        alerts=alert_items,
        checked_at=latest_checked_at,
    )


@router.post("/{project_id}/health/refresh", response_model=HealthCheckResp)
async def refresh_project_health(
    project_id: int,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
) -> HealthCheckResp:
    """刷新项目健康告警。

    运行所有健康检查（R1/R2/R3）并返回最新的告警列表。

    Args:
        project_id: 项目 ID
        db: 数据库会话
        current_user: 当前认证用户

    Returns:
        HealthCheckResp: 包含最新告警列表和检查时间
    """
    # 运行健康检查
    result = await health_service.run_health_check(
        db, str(current_user.id), project_id
    )

    # 转换为响应格式
    alert_items = []
    for alert in result.get("alerts", []):
        alert_items.append(HealthAlertItem(
            rule=alert.get("rule", ""),
            title=alert.get("title", ""),
            detail=alert.get("detail", ""),
        ))

    checked_at = datetime.now(timezone.utc)
    if result.get("checked_at"):
        try:
            checked_at = datetime.fromisoformat(result["checked_at"].replace("Z", "+00:00"))
        except (ValueError, TypeError):
            pass

    return HealthCheckResp(
        alerts=alert_items,
        checked_at=checked_at,
    )

File: moling-server-rs/crates/moling-api/src/routes/project_health.rs
//! Project health routes — R1/R2/R3 health check alerts.
//!
//! Ported from Python `app/router/project_health.py`.
//!
//! Endpoints:
//! - GET  /{project_id}/health         — get active health alerts
//! - POST /{project_id}/health/refresh — run health checks and return results
//!
//! Uses [`moling_services::HealthService`] for alert management and
//! health check execution.

use axum::{
    extract::{Path, State},
    routing::{get, post},
    Json, Router,
};
use moling_core::error::AppResult;
use moling_services::HealthService;
use moling_auth::CurrentUser;
use crate::state::AppState;
use crate::types::{HealthAlertItem, HealthCheckResp};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/{project_id}/health", get(get_project_health))
        .route("/{project_id}/health/refresh", post(refresh_project_health))
}

// ======================================================================
// GET /{project_id}/health
// ======================================================================

/// Get active health alerts for a project.
async fn get_project_health(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(project_id): Path<i32>,
) -> AppResult<Json<HealthCheckResp>> {
    let health = HealthService::new();
    let alerts = health.list_alerts(&state.db, project_id, true).await?;

    let alert_items: Vec<HealthAlertItem> = alerts
        .iter()
        .map(|a| HealthAlertItem {
            rule: a.rule.clone(),
            title: a.title.clone(),
            detail: a.detail.clone(),
            severity: Some(a.severity.clone()),
            level: None,
            is_active: Some(a.is_active),
        })
        .collect();

    let latest_checked_at = alerts
        .iter()
        .filter_map(|a| a.checked_at)
        .max()
        .map(|t| t.to_rfc3339())
        .unwrap_or_else(|| chrono::Utc::now().to_rfc3339());

    let _ = user; // ownership verified via project access implicitly

    Ok(Json(HealthCheckResp {
        alerts: alert_items,
        checked_at: latest_checked_at,
        status: None,
        alert_counts: None,
    }))
}

// ======================================================================
// POST /{project_id}/health/refresh
// ======================================================================

/// Run all health checks (R1/R2/R3) and return updated alerts.
async fn refresh_project_health(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(project_id): Path<i32>,
) -> AppResult<Json<HealthCheckResp>> {
    let health = HealthService::new();
    let user_id = user.user_id.to_string();

    let check_result = health.run_health_check(&state.db, &user_id, project_id).await?;

    // Extract alert items from the check result
    let mut alert_items: Vec<HealthAlertItem> = Vec::new();
    if let Some(alerts) = check_result.get("alerts").and_then(|v| v.as_array()) {
        for alert in alerts {
            alert_items.push(HealthAlertItem {
                rule: alert.get("rule").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                title: alert.get("title").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                detail: alert.get("detail").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                severity: alert.get("severity").and_then(|v| v.as_str()).map(String::from),
                level: None,
                is_active: None,
            });
        }
    }

    let checked_at = chrono::Utc::now().to_rfc3339();

    Ok(Json(HealthCheckResp {
        alerts: alert_items,
        checked_at,
        status: None,
        alert_counts: None,
    }))
}
